### PR TITLE
 Memory-safety proofs for aws_byte_buf 

### PR DESCRIPTION
**Pull request by feliperodri**
*Description of changes:*

Adds preconditions, postconditions and proof harnesses for the following functions:
 - aws_byte_buf_append
 - aws_byte_buf_append_dynamic
 - aws_byte_buf_append_with_lookup
 - aws_byte_buf_clean_up
 - aws_byte_buf_from_array
 - aws_byte_buf_from_empty_array
 - aws_byte_buf_init
 - aws_byte_buf_init_copy_from_cursor
 - aws_byte_buf_reserve
 - aws_byte_buf_secure_zero

Updates the implementation of the following functions to correctly handle empty `aws_byte_buf` structures:
 - aws_byte_buf_from_array
 - aws_byte_buf_from_empty_array
 - aws_byte_buf_init
 - aws_byte_buf_init_copy
 - aws_byte_buf_is_valid
 - aws_byte_cursor_is_valid
 - aws_byte_buf_init_copy_from_cursor
 - aws_byte_buf_append_dynamic
 - aws_byte_buf_reserve
 - aws_byte_buf_reserve_relative
 - aws_hex_encode
 - aws_hex_decode

Updates the following test cases:
 - test_buffer_init_copy_null_buffer
 - s_run_hex_encoding_test_case

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by danielsn** [comment]
The upstream issues will be fixed by https://github.com/awslabs/aws-c-io/pull/132

File: source/byte_buf.c [hunk 1]
```diff
@@ -53,26 +66,17 @@ int aws_byte_buf_init_copy(struct aws_byte_buf *dest, struct aws_allocator *allo
 }
 
 bool aws_byte_buf_is_valid(const struct aws_byte_buf *const buf) {
```
Comment (feliperodri) [inline source/byte_buf.c]: The new invariant assumes an empty `aws_byte_buf` is valid, any comments @justinboswell @bretambrose @danielsn ?

File: source/byte_buf.c [hunk 2]
```diff
@@ -53,26 +66,17 @@ int aws_byte_buf_init_copy(struct aws_byte_buf *dest, struct aws_allocator *allo
...
-    }
-    bool buffer_is_valid = (buf->buffer && AWS_MEM_IS_WRITABLE(buf->buffer, buf->len));
-    bool capacity_is_valid = (buf->capacity > 0);
-    bool len_is_valid = (buf->len <= buf->capacity);
-    return capacity_is_valid && buffer_is_valid && len_is_valid;
+    return buf && ((buf->buffer == NULL && buf->capacity == 0 && buf->len == 0) ||
+                   (buf->capacity > 0 && buf->len <= buf->capacity && AWS_MEM_IS_WRITABLE(buf->buffer, buf->len)));
 }
 
 bool aws_byte_cursor_is_valid(const struct aws_byte_cursor *cursor) {
```
Comment (feliperodri) [inline source/byte_buf.c]: The new invariant assumes a 0 length `aws_byte_cursor` is valid always valid, any comments @justinboswell @bretambrose @danielsn ?

File: .cbmc-batch/jobs/Makefile.aws_byte_buf [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+#
+# or in the "license" file accompanying this file. This file is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+##########
+
+# Sufficently long to get full coverage on the aws_byte_buf and aws_byte_cursor APIs
+# short enough that all proofs complete quickly
```
Comment (danielsn) [inline .cbmc-batch/jobs/Makefile.aws_byte_buf]: Approx how quickly (order of magnitude).

File: .cbmc-batch/jobs/aws_byte_buf_append/aws_byte_buf_append_harness.c [hunk 4]
```diff
@@ -17,22 +17,20 @@
...
 
 void aws_byte_buf_append_harness() {
-    size_t len1 = nondet_size_t();
-    __CPROVER_assume(len1 <= MAX_BUF_LEN);
-    size_t len2 = nondet_size_t();
+    struct aws_byte_buf to;
+    __CPROVER_assume(aws_byte_buf_is_bounded(&to, MAX_BUFFER_SIZE));
+    ensure_byte_buf_has_allocated_buffer_member(&to);
+    __CPROVER_assume(aws_byte_buf_is_valid(&to));
+    __CPROVER_assume(to.buffer != NULL);
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_byte_buf_append/aws_byte_buf_append_harness.c]: Is this required? What happens in the 0/0 case?

File: .cbmc-batch/jobs/aws_byte_buf_append/aws_byte_buf_append_harness.c [hunk 5]
```diff
@@ -17,22 +17,20 @@
...
-    struct aws_byte_buf *to;
-    ASSUME_VALID_MEMORY(to);
-    ASSUME_VALID_MEMORY_COUNT(to->buffer, len1);
-    to->capacity = len1;
-    __CPROVER_assume(to->len <= to->capacity);
+    struct aws_byte_cursor from;
+    __CPROVER_assume(aws_byte_cursor_is_bounded(&from, MAX_BUFFER_SIZE));
+    ensure_byte_cursor_has_allocated_buffer_member(&from);
+    __CPROVER_assume(aws_byte_cursor_is_valid(&from));
+    __CPROVER_assume(from.ptr != NULL);
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_byte_buf_append/aws_byte_buf_append_harness.c]: What about the 0/0 case?

File: .cbmc-batch/jobs/aws_byte_buf_append_with_lookup/aws_byte_buf_append_with_lookup_harness.c [hunk 6]
```diff
@@ -35,6 +35,5 @@ void aws_byte_buf_append_with_lookup_harness() {
     aws_byte_buf_append_with_lookup(&to, &from, lookup_table);
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_byte_buf_append_with_lookup/aws_byte_buf_append_with_lookup_harness.c]: This returns an error code: are there things to check on success/failure?

File: .cbmc-batch/jobs/aws_byte_buf_append/aws_byte_buf_append_harness.c [hunk 7]
```diff
@@ -17,22 +17,20 @@
...
 
-    /* Need arbitrary cursor */
-    struct aws_byte_cursor *from;
-    ASSUME_VALID_MEMORY(from);
-    from->ptr = malloc(len2);
-    __CPROVER_assume(from->len <= len2);
+    aws_byte_buf_append(&to, &from);
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_byte_buf_append/aws_byte_buf_append_harness.c]: This returns an error code: are there things to check on success/failure?

File: .cbmc-batch/jobs/aws_byte_buf_init_copy_from_cursor/aws_byte_buf_init_copy_from_cursor_harness.c [hunk 8]
```diff
@@ -26,20 +26,21 @@ void aws_byte_buf_init_copy_from_cursor_harness() {
     struct aws_byte_cursor cursor;
 
     /* assumptions */
-    __CPROVER_assume(is_bounded_byte_cursor(&cursor, MAX_BUFFER_SIZE));
+    __CPROVER_assume(aws_byte_cursor_is_bounded(&cursor, MAX_BUFFER_SIZE));
     ensure_byte_cursor_has_allocated_buffer_member(&cursor);
-    __CPROVER_assume(aws_byte_cursor_is_valid(&cursor));
+    //__CPROVER_assume(aws_byte_cursor_is_valid(&cursor));
     ASSUME_VALID_MEMORY(buf);
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_byte_buf_init_copy_from_cursor/aws_byte_buf_init_copy_from_cursor_harness.c]: Just make buf declared on the stack, and the need for this goes away

File: .cbmc-batch/jobs/aws_byte_buf_reserve/aws_byte_buf_reserve_harness.c [hunk 9]
```diff
@@ -23,11 +23,11 @@ void aws_byte_buf_reserve_harness() {
 
     struct aws_byte_buf old = buf;
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_byte_buf_reserve/aws_byte_buf_reserve_harness.c]: Do you ever do anything with this?

File: include/aws/common/byte_buf.h [hunk 10]
```diff
@@ -459,19 +459,21 @@ AWS_STATIC_IMPL struct aws_byte_buf aws_byte_buf_from_c_str(const char *c_str) {
 
 AWS_STATIC_IMPL struct aws_byte_buf aws_byte_buf_from_array(const void *bytes, size_t len) {
     struct aws_byte_buf buf;
-    buf.buffer = (uint8_t *)bytes;
+    buf.buffer = (len) ? (uint8_t *)bytes : NULL;
```
Comment (danielsn) [inline include/aws/common/byte_buf.h]: I personally find ```len > 0``` clearer here.

File: include/aws/common/byte_buf.h [hunk 11]
```diff
@@ -459,19 +459,21 @@ AWS_STATIC_IMPL struct aws_byte_buf aws_byte_buf_from_c_str(const char *c_str) {
 
 AWS_STATIC_IMPL struct aws_byte_buf aws_byte_buf_from_array(const void *bytes, size_t len) {
     struct aws_byte_buf buf;
```
Comment (danielsn) [inline include/aws/common/byte_buf.h]: precondition that ```AWS_MEMORY_IS_VALID(bytes,len)```

File: include/aws/common/byte_buf.h [hunk 12]
```diff
@@ -459,19 +459,21 @@ AWS_STATIC_IMPL struct aws_byte_buf aws_byte_buf_from_c_str(const char *c_str) {
...
-    buf.buffer = (uint8_t *)bytes;
+    buf.buffer = (len) ? (uint8_t *)bytes : NULL;
     buf.len = len;
     buf.capacity = len;
     buf.allocator = NULL;
+    AWS_POSTCONDITION(aws_byte_buf_is_valid(&buf));
     return buf;
 }
 
 AWS_STATIC_IMPL struct aws_byte_buf aws_byte_buf_from_empty_array(const void *bytes, size_t capacity) {
```
Comment (danielsn) [inline include/aws/common/byte_buf.h]: precondition that AWS_MEMORY_IS_VALID(bytes,len)

File: source/byte_buf.c [hunk 13]
```diff
@@ -25,13 +25,19 @@
 #endif
 
 int aws_byte_buf_init(struct aws_byte_buf *buf, struct aws_allocator *allocator, size_t capacity) {
-    buf->buffer = (uint8_t *)aws_mem_acquire(allocator, capacity);
-    if (!buf->buffer) {
+    if (!buf || !allocator) {
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+
+    buf->buffer = capacity == 0 ? NULL : aws_mem_acquire(allocator, capacity);
```
Comment (danielsn) [inline source/byte_buf.c]: ```(capacity == 0)```

File: source/byte_buf.c [hunk 14]
```diff
@@ -405,28 +413,39 @@ bool aws_byte_cursor_eq_c_str_ignore_case(const struct aws_byte_cursor *cursor,
 }
 
 int aws_byte_buf_append(struct aws_byte_buf *to, const struct aws_byte_cursor *from) {
-    assert(from->ptr);
-    assert(to->buffer);
+    AWS_PRECONDITION(aws_byte_buf_is_valid(to));
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(from));
+    if (!from->ptr || !to->buffer) {
```
Comment (danielsn) [inline source/byte_buf.c]: do we need this on top of the existing precondtions

File: source/byte_buf.c [hunk 15]
```diff
@@ -405,28 +413,39 @@ bool aws_byte_cursor_eq_c_str_ignore_case(const struct aws_byte_cursor *cursor,
...
 
 int aws_byte_buf_append_with_lookup(
     struct aws_byte_buf *AWS_RESTRICT to,
     const struct aws_byte_cursor *AWS_RESTRICT from,
     const uint8_t *lookup_table) {
-    AWS_PRECONDITION(from->ptr);
     AWS_PRECONDITION(aws_byte_buf_is_valid(to));
     AWS_PRECONDITION(aws_byte_cursor_is_valid(from));
     AWS_PRECONDITION(AWS_MEM_IS_READABLE(lookup_table, 256));
+    if (!from->ptr || !to->buffer) {
```
Comment (danielsn) [inline source/byte_buf.c]: Same here: is this additional to the other preconditions.

File: tests/encoding_test.c [hunk 16]
```diff
@@ -71,7 +71,11 @@ static int s_run_hex_encoding_test_case(
     ASSERT_INT_EQUALS(test_str_size - 1, output_size, "Output size on string should be %d", test_str_size - 1);
 
     output.capacity = output_size;
+    if (output.capacity == 0) {
```
Comment (danielsn) [inline tests/encoding_test.c]: This kind of poking into the internals of a buffer feels fishy to me.

File: source/byte_buf.c (hunk 14 above)
Comment (feliperodri) [reply source/byte_buf.c]: Yes, because we consider empty `buf` or `cursor` as valid. However, `aws_byte_buf_append` does not work on empty structures. Of course, we could also update the implementation to consider them. What would you suggest?

File: source/byte_buf.c (hunk 15 above)
Comment (feliperodri) [reply source/byte_buf.c]: Same answer too.

**Comment by feliperodri** [comment]
@karkhaz any feedback on the harnesses?

File: .cbmc-batch/jobs/aws_byte_buf_clean_up/Makefile [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+#
+# or in the "license" file accompanying this file. This file is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+###########
+include ../Makefile.aws_byte_buf
+
+UNWINDSET += memset_impl.0:41
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_byte_buf_clean_up/Makefile]: Calculate this from the MAX_BUFFER_SIZE

File: .cbmc-batch/jobs/aws_byte_buf_from_array/aws_byte_buf_from_array_harness.c [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+
+    /* assumptions */
+    ASSUME_VALID_MEMORY_COUNT(array, length);
+
+    /* operation under verification */
+    struct aws_byte_buf buf = aws_byte_buf_from_array(array, length);
+
+    /* assertions */
+    assert(aws_byte_buf_is_valid(&buf));
+    size_t debug_len = buf.len;
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_byte_buf_from_array/aws_byte_buf_from_array_harness.c]: Why is this here

File: .cbmc-batch/jobs/aws_byte_buf_init/aws_byte_buf_init_harness.c [hunk 19]
```diff
@@ -18,12 +18,27 @@
...
     struct aws_byte_buf *buf;
-    ASSUME_VALID_MEMORY(buf);
 
-    size_t len = nondet_size_t();
+    /* parameters */
+    struct aws_allocator *allocator;
+    size_t capacity;
+
+    /* assumptions */
+    ASSUME_VALID_MEMORY(buf);
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_byte_buf_init/aws_byte_buf_init_harness.c]: Why not just declare buf on the stack

File: .cbmc-batch/jobs/aws_byte_buf_reserve_relative/aws_byte_buf_reserve_relative_harness.c [hunk 20]
```diff
@@ -27,7 +27,7 @@ void aws_byte_buf_reserve_relative_harness() {
 
     if (rval == AWS_OP_SUCCESS) {
         assert(buf.capacity >= (old.len + requested_capacity));
+        assert(is_byte_buf_expected_alloc(&buf));
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_byte_buf_reserve_relative/aws_byte_buf_reserve_relative_harness.c]: Maybe rename this function to follow the naming scheme?

File: .cbmc-batch/jobs/aws_byte_buf_secure_zero/Makefile [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+#
+# or in the "license" file accompanying this file. This file is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+###########
+include ../Makefile.aws_byte_buf
+
+UNWINDSET += memset_impl.0:41
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_byte_buf_secure_zero/Makefile]: Calculate this number

File: source/byte_buf.c [hunk 22]
```diff
@@ -53,26 +66,17 @@ int aws_byte_buf_init_copy(struct aws_byte_buf *dest, struct aws_allocator *allo
...
 
 bool aws_byte_buf_is_valid(const struct aws_byte_buf *const buf) {
-    if (!buf) {
-        return false;
-    }
-    bool buffer_is_valid = (buf->buffer && AWS_MEM_IS_WRITABLE(buf->buffer, buf->len));
-    bool capacity_is_valid = (buf->capacity > 0);
-    bool len_is_valid = (buf->len <= buf->capacity);
-    return capacity_is_valid && buffer_is_valid && len_is_valid;
+    return buf && ((buf->buffer == NULL && buf->capacity == 0 && buf->len == 0) ||
```
Comment (danielsn) [inline source/byte_buf.c]: Put the conditions in the same order in both branches.

File: source/byte_buf.c [hunk 23]
```diff
@@ -53,26 +66,17 @@ int aws_byte_buf_init_copy(struct aws_byte_buf *dest, struct aws_allocator *allo
...
 bool aws_byte_cursor_is_valid(const struct aws_byte_cursor *cursor) {
-    bool ptr_is_valid;
-    if (cursor->ptr) {
-        ptr_is_valid = (cursor->len && AWS_MEM_IS_WRITABLE(cursor->ptr, cursor->len));
-    } else {
-        ptr_is_valid = (!cursor->len);
-    }
-    return ptr_is_valid;
+    return cursor &&
+           ((cursor->len == 0) || (cursor->ptr && cursor->len > 0 && AWS_MEM_IS_WRITABLE(cursor->ptr, cursor->len)));
```
Comment (danielsn) [inline source/byte_buf.c]: conditions in same order.

File: source/byte_buf.c [hunk 24]
```diff
@@ -118,21 +124,23 @@ int aws_byte_buf_init_copy_from_cursor(
...
+    dest->buffer = (src.len == 0) ? NULL : (uint8_t *)aws_mem_acquire(allocator, src.len);
+    if (src.len != 0 && dest->buffer == NULL) {
         return AWS_OP_ERR;
     }
 
     dest->len = src.len;
     dest->capacity = src.len;
     dest->allocator = allocator;
-    memcpy(dest->buffer, src.ptr, src.len);
+    if (dest->buffer) {
```
Comment (danielsn) [inline source/byte_buf.c]: assert.  If ```dest->buffer``` is null, we have an error, right?

File: source/byte_buf.c [hunk 25]
```diff
@@ -405,28 +413,39 @@ bool aws_byte_cursor_eq_c_str_ignore_case(const struct aws_byte_cursor *cursor,
...
     if (to->capacity - to->len < from->len) {
+        AWS_POSTCONDITION(aws_byte_buf_is_valid(to));
+        AWS_POSTCONDITION(aws_byte_cursor_is_valid(from));
         return aws_raise_error(AWS_ERROR_DEST_COPY_TOO_SMALL);
     }
 
-    memcpy(to->buffer + to->len, from->ptr, from->len);
-    to->len += from->len;
+    if (from->len > 0) {
+        /* This assert teaches clang-tidy and friends that from->ptr and to->buffer cannot be null in a non-empty */
```
Comment (danielsn) [inline source/byte_buf.c]: In a non-empty what?

File: source/byte_buf.c [hunk 26]
```diff
@@ -444,18 +463,22 @@ int aws_byte_buf_append_with_lookup(
...
 int aws_byte_buf_append_dynamic(struct aws_byte_buf *to, const struct aws_byte_cursor *from) {
-    AWS_PRECONDITION(from->ptr);
     AWS_PRECONDITION(aws_byte_buf_is_valid(to));
     AWS_PRECONDITION(aws_byte_cursor_is_valid(from));
+    if (!to->allocator) {
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
 
-    if (to->capacity - to->len < from->len) {
+    if (to->capacity - to->len < from->len || to->capacity == 0) {
```
Comment (danielsn) [inline source/byte_buf.c]: why is the ```to->capacity == 0``` necessary?

File: source/byte_buf.c [hunk 27]
```diff
@@ -513,7 +542,12 @@ int aws_byte_buf_append_dynamic(struct aws_byte_buf *to, const struct aws_byte_c
         to->buffer = new_buffer;
         to->capacity = new_capacity;
     } else {
-        memcpy(to->buffer + to->len, from->ptr, from->len);
+        if (from->len > 0) {
+            /* This assert teaches clang-tidy and friends that from->ptr and to->buffer cannot be null in a non-empty */
```
Comment (danielsn) [inline source/byte_buf.c]: non-empty what?

File: source/byte_buf.c [hunk 28]
```diff
@@ -53,26 +66,17 @@ int aws_byte_buf_init_copy(struct aws_byte_buf *dest, struct aws_allocator *allo
...
-    }
-    bool buffer_is_valid = (buf->buffer && AWS_MEM_IS_WRITABLE(buf->buffer, buf->len));
-    bool capacity_is_valid = (buf->capacity > 0);
-    bool len_is_valid = (buf->len <= buf->capacity);
-    return capacity_is_valid && buffer_is_valid && len_is_valid;
+    return buf && ((buf->capacity == 0 && buf->len == 0 && buf->buffer == NULL) ||
+                   (buf->capacity > 0 && buf->len <= buf->capacity && AWS_MEM_IS_WRITABLE(buf->buffer, buf->len)));
 }
 
 bool aws_byte_cursor_is_valid(const struct aws_byte_cursor *cursor) {
```
Comment (feliperodri) [inline source/byte_buf.c]: The new invariant assumes a 0 length `aws_byte_cursor` is valid always valid, any comments @justinboswell @bretambrose?

File: source/byte_buf.c (hunk 1 above)
Comment (feliperodri) [inline source/byte_buf.c]: The new invariant assumes an empty `aws_byte_buf` is valid, any comments @justinboswell @bretambrose?

Comment (bretambrose) [reply source/byte_buf.c]: Looks good

File: source/byte_buf.c (hunk 28 above)
Comment (bretambrose) [reply source/byte_buf.c]: I'm happy!
